tracing: propagate non-recording spans across rpc boundaries

tbg opened the pull request:
We were previously propagating RPCs across boundaries only when they
were verbose. We now propagate any span (except the noop span)
regardless of current verbosity.
This ensures that SQL (which always creates real, non-verbose spans by
default) can trace the entirety of its operations.

This enables the background collection of contention metadata in 21.1
and more generally, the idea of always-on tracing. Note that we had
previously considered always-on tracing to mean avoiding the noop span
altogether; due to memory retention issues[1][2] encountered we are
taking a step back from that idea for the 21.1 release.

[1]: https://github.com/cockroachdb/cockroach/pull/59370
[2]: https://github.com/cockroachdb/cockroach/pull/59431

Release note: None

tbg committed:
tracing: propagate non-recording spans across rpc boundaries

We were previously propagating RPCs across boundaries only when they
were verbose. We now propagate any span (except the noop span)
regardless of current verbosity.
This ensures that SQL (which always creates real, non-verbose spans by
default) can trace the entirety of its operations.

This enables the background collection of contention metadata in 21.1
and more generally, the idea of always-on tracing. Note that we had
previously considered always-on tracing to mean avoiding the noop span
altogether; due to memory retention issues[1][2] encountered we are
taking a step back from that idea for the 21.1 release.

Somehow creating additional trace spans seems to have fouled up some
tracing-based tests. I can't boil that ocean here, so I filed a
follow-up issue: #60672

[1]: https://github.com/cockroachdb/cockroach/pull/59370
[2]: https://github.com/cockroachdb/cockroach/pull/59431

Release note: None

tbg committed:
logictest: improve contention_event

Switch to the 5node family of clusters to test against (more realistic),
and also make the query that collects the event touch multiple ranges,
which make it "easier" to miss an event.

cc @irfansharif - this could help with the discussion about DistSender
in [1].

[1]: https://github.com/cockroachdb/cockroach/pull/59815

Release note: None

tbg committed:
acceptance: enable tracing in acceptance/version-upgrade

If we introduced NPEs due to incompatibilities between versions, we
would want to find out. Enabling tracing globally while upgrading
clusters should help.

Release note: None

tbg committed:
kvnemeses: enable global tracing in KVNemesisMultiNode

Enabling tracing gives this test a chance to hit random NPEs.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59992)
<!-- Reviewable:end -->

irfansharif reviewed (approved):
Stray "iasd" in the first commit message, typo with "Mult" in second commit.

> Reviewed 6 of 6 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-:-MT1ngP7AjAIQEnWquz_:b-os46y5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @knz, and @tbg)

___
*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT1pbkz0mxbvJP7HPQL:-MT1pbkz0mxbvJP7HPQM:b-85bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b33e6dfc47000de831745a851e3bf9e2cf7fd95/pkg/sql/logictest/testdata/logic_test/distsql_agg#L1)):*
> ```
> # LogicTest: 5node-default-configs
> ```

Why did this change?
___
*[pkg/util/tracing/span_test.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT1oTIe9XCahoa3knFc:-MT1oTIe9XCahoa3knFd:b-85bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b33e6dfc47000de831745a851e3bf9e2cf7fd95/pkg/util/tracing/span_test.go#L187)):*
> ```Go
> func TestSpan_LogStructured(t *testing.T) {
> 	tr := NewTracer()
> 	tr._mode = int32(modeBackground)
> ```

Why did this change?
___
*[pkg/util/tracing/tracer.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT1oeKdEvdx9rhZ48bp:-MT1oeKeChZxWkGCToor:b-a8yf5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9cbe5776de09a2f7b38b73c32987e327d60dadb/pkg/util/tracing/tracer.go#L297)):*
> ```Go
> 	// a real span they need to get one. In all other cases, a noop span will do.
> 	if !t.AlwaysTrace() &&
> 		opts.parentTraceID() == 0 &&
> ```

I like this simplification. The semantics around recording being switched on/off and what that means for log messages/structured events still feels a bit confused to me (and was trying something elsewhere), and this makes it a bit easier to reason about.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-:-MT55jG07jV7Pk3iC7Ye:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, and @knz)

___
*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT1pbkz0mxbvJP7HPQL:-MT559eHEY1W9hJR_kMN:b-ckg5ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b33e6dfc47000de831745a851e3bf9e2cf7fd95/pkg/sql/logictest/testdata/logic_test/distsql_agg#L1)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Why did this change?
</blockquote></details>

Thanks for catching that, this is accidental, fixed.
___
*[pkg/util/tracing/span_test.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT1oTIe9XCahoa3knFc:-MT55Mb68FL_vr_WNxtd:b9p2j9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b33e6dfc47000de831745a851e3bf9e2cf7fd95/pkg/util/tracing/span_test.go#L187)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Why did this change?
</blockquote></details>

To validate the changes in `GetRecording`. Without the fixes to it, this test would have started failing as we short-circuited `GetRecording()` to nil.


<!-- Sent from Reviewable.io -->

tbg commented:
```
I210209 11:49:00.327146 1 util/log/flags.go:106  stderr capture started
panic: interface conversion: opentracing.Span is *tracing.noopSpan, not *tracing.span [recovered]
	panic: interface conversion: opentracing.Span is *tracing.noopSpan, not *tracing.span

goroutine 628 [running]:
panic(0x4bb5760, 0xc0017c3b30)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc001737f80 sp=0xc001737ef0 pc=0x7b1a62
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0004dad80, 0x63133c0, 0xc0017e2ae0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:207 +0x11f fp=0xc001737fe0 sp=0xc001737f80 pc=0x169df4f
runtime.call32(0x0, 0x5808980, 0xc00173bcf8, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:539 +0x3b fp=0xc001738010 sp=0xc001737fe0 pc=0x7e051b
panic(0x4bb5760, 0xc0017c3b30)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2 fp=0xc0017380a0 sp=0xc001738010 pc=0x7b1952
runtime.panicdottypeE(...)
	/usr/local/go/src/runtime/iface.go:255
runtime.panicdottypeI(0x647e040, 0x510f8e0, 0x4ed8d80)
	/usr/local/go/src/runtime/iface.go:265 +0xf4 fp=0xc0017380c8 sp=0xc0017380a0 pc=0x78c3e4
github.com/cockroachdb/cockroach/pkg/util/tracing.ImportRemoteSpans(0x647e040, 0xc000b6a000, 0xc000171000, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/tracer_span.go:638 +0x290 fp=0xc001738138 sp=0xc0017380c8 pc=0x111d820
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch(0xc0015f4840, 0x63133c0, 0xc0017e2bd0, 0x3, 0x633f100, 0xc0008940b0, 0x166211f05167de20, 0x0, 0x300000003, 0x3, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:173 +0x390 fp=0xc0017381f0 sp=0xc001738138 pc=0x1bd2100
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext(0xc0015f4840, 0x63133c0, 0xc0017e2bd0, 0x166211f05167de20, 0x0, 0x300000003, 0x3, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:139 +0x186 fp=0xc0017382e8 sp=0xc0017381f0 pc=0x1bd1d16
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas(0xc000865000, 0x63133c0, 0xc0017e2bd0, 0x166211f05167de20, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1870 +0x7d0 fp=0xc001738d80 sp=0xc0017382e8 pc=0x1bc43f0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc000865000, 0x63133c0, 0xc0017e2bd0, 0x166211f05167de20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1495 +0x346 fp=0xc00173a1a8 sp=0xc001738d80 pc=0x1bc0946
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc000865000, 0x63133c0, 0xc0017e2bd0, 0x166211f05167de20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1132 +0x16f4 fp=0xc00173a808 sp=0xc00173a1a8 pc=0x1bbfe14
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc000865000, 0x63133c0, 0xc0017e2bd0, 0x166211f05167de20, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:771 +0x7a6 fp=0xc00173ab68 sp=0xc00173a808 pc=0x1bbc8b6
github.com/cockroachdb/cockroach/pkg/kv.lookupRangeFwdScan(0x63133c0, 0xc0017e2bd0, 0x626a800, 0xc000865000, 0xc0017c4ef0, 0xe, 0xe, 0x1, 0x8, 0x1dcd6500, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_lookup.go:312 +0x42b fp=0xc00173af30 sp=0xc00173ab68 pc=0x180baab
github.com/cockroachdb/cockroach/pkg/kv.RangeLookup(0x63133c0, 0xc0017e2bd0, 0x626a800, 0xc000865000, 0xc0017c4ef0, 0xe, 0xe, 0x1, 0x8, 0xab7400, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_lookup.go:197 +0x7d2 fp=0xc00173b2e0 sp=0xc00173af30 pc=0x180b1e2
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).RangeLookup(0xc000865000, 0x63133c0, 0xc0017e2bd0, 0xc0017c4ef0, 0xe, 0xe, 0x6313300, 0xc0017e2bd0, 0x977c24, 0xc000508d90, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:447 +0xd4 fp=0xc00173b380 sp=0xc00173b2e0 pc=0x1bbace4
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*RangeDescriptorCache).performRangeLookup(0xc000f4e3c0, 0x63133c0, 0xc0017e2bd0, 0xc0017c4ef0, 0xe, 0xe, 0x0, 0xc0005834d8, 0x97825b, 0x63133c0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/range_cache.go:818 +0x285 fp=0xc00173b450 sp=0xc00173b380 pc=0x1bcc7b5
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*RangeDescriptorCache).tryLookup.func1.1.1(0x6313380, 0xc000508d80, 0x2540be400, 0x6313380)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/range_cache.go:672 +0x8e fp=0xc00173b4e8 sp=0xc00173b450 pc=0x1be6a7e
github.com/cockroachdb/cockroach/pkg/util/contextutil.RunWithTimeout(0x6313380, 0xc000508d80, 0x521601f, 0xc, 0x2540be400, 0xc000583730, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/contextutil/context.go:135 +0xbf fp=0xc00173b570 sp=0xc00173b4e8 pc=0x16a12cf
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*RangeDescriptorCache).tryLookup.func1.1(0x63133c0, 0xc0017e2ae0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/range_cache.go:669 +0x313 fp=0xc00173bc60 sp=0xc00173b570 pc=0x1be6e53
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0004dad80, 0x63133c0, 0xc0017e2ae0, 0x524387e, 0x18, 0xc00152fd60, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x140 fp=0xc00173bd20 sp=0xc00173bc60 pc=0x169e8f0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*RangeDescriptorCache).tryLookup.func1(0x16a03b4, 0x580c5e8, 0x0, 0xc00152ff70)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/range_cache.go:657 +0xe9 fp=0xc00173bf48 sp=0xc00173bd20 pc=0x1be7859
github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall(0xc000f4e400, 0xc000b00e10, 0xc000076920, 0x15, 0xc0015f43c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:128 +0x2e fp=0xc00173bfb8 sp=0xc00173bf48 pc=0x1bb14ae
runtime.goexit()
```

Oh wonderful I love the smell of mixed version problems in the afternoon. Will take a bit longer to get this in I suppose, this is crashing the *old* version not the new.

tbg commented:
Wow so I'm surprised we haven't seen crashes here more (maybe the crash here isn't even this PR, maybe it's just crashing now because I just added a step to turn tracing on in this test. The code on 20.2.4 is pretty sensitive to crashes because it has these assertions that make no sense, in the methods called from this area:

https://github.com/cockroachdb/cockroach/blob/9c7ef622f9c37ae423d2b917896e7ea226058c5f/pkg/kv/kvclient/kvcoord/transport.go#L160-L179

Concretely, here what we see is this:

- we crash because we call `ImportRemoteSpans` on the noop span: https://github.com/cockroachdb/cockroach/blob/a9eb3faa0210eb583664a63a913d64e0ce88f004/pkg/util/tracing/tracer_span.go#L638
- we call ImportRemoteSpans indiscriminately, whenever a batch response contains a collected trace (see the snippet above)

I assume that the way this particular crash happens is this:

- test sets tracing to on, setting gets pushed out via gossip
- n1 sends an RPC to n2. n1 isn't tracing the request yet, so the span is a noop span.
- when the request arrives at n2, n2 inspects it and sees that there is no trace attached. Since tracing is globally on, it creates a new trace root span and returns the collected recording with the response
- n1 explodes trying to shove that response into the noop span

But there could also have been a different scenario (pre this PR), though not one that you could hit in practice, since we don't use spans in that way at that part of the stack:

- n1 has a real span, but it's not recording
- consequently, it attaches no trace metadata to the RPC
- n2 is tracing, so it makes a new trace (different traceid) and returns recording
- n1 hits this error: https://github.com/cockroachdb/cockroach/blob/a9eb3faa0210eb583664a63a913d64e0ce88f004/pkg/util/tracing/tracer_span.go#L639-L641

Now in 20.2.4, if you don't mess with the global tracing (which you shouldn't anyway, as it might well OOM you), you don't see any of this because nobody ever sents any recordings back. But with this PR, since SQL on the new nodes will *always* propagate a real trace, oof, we're exercising this code path all of the time.

At this point, and seeing how we've made more changes to the tracer than ever before including moving off the OT interfaces, I think we'll play it safe - I'll add a cluster version and make sure that new nodes don't create any spans, period (unless the verbose cluster setting is set, in which case, good luck). I'll also give the code around the interfaces a do-over to make sure they are not exceedingly... assertive. Something that should happen in that context is to gracefully handle importing a recording for trace B into trace A. This should silently ignore the mismatched data. Of course we don't expect that to happen (after this PR), but resilience here will pay off and is cheap.

tbg commented:
> acceptance: enable tracing in acceptance/version-upgrade

With only this commit, `acceptance/version-upgrade` passed 6/6 runs. This makes sense, since the test enables tracing before going to upload a binary, so there usually isn't anything going on on the cluster that would have a high chance of triggering https://github.com/cockroachdb/cockroach/pull/59992#issuecomment-778146715. It's only with the main change here that recordings not expected by v20.2 come back and cause the panic.

tbg committed:
tracing: fix accidental noop in ImportRemoteSpans

There was vestigial code that would drop a received recording on the
floor if the receiving Span wasn't verbose. We now want to ingest
Recordings independently of that setting.

Release note: None

tbg committed:
tracing: prevent sending unexpected recordings to v20.2 nodes

Unfortunately, v20.2 fails when it is presented with a recording (from a
remote node) that it did not expect to get created. Concretely, we can
end up with a noopSpan having `ImportRemoteSpans` called on it, which
is a fatal error (unhandled failed interface assertion).

To avoid this particular problem and any others that may lurk around,
simply never send recordings back that would not be created under
v20.2, until we know there is no v20.2 node around any more. This
is a straightforward introduction of our cluster version, with the
exception that a little bit of dependency injection is necessary to
avoid circular imports.

Release note: None

tbg committed:
sql: small but essential tracing fixes

1. Require admin role for `crdb_internal.node_inflight_trace_spans`
2. Add num_structured column to that table (this can be seen as
   temporary for testing; we can remove it once #55733 is fully
   addressed and payloads can be introspected separately).
3. Add workaround for failure to finish sql txn spans (#59315)
   for use in logic tests, which together with 2.) and 4.) enables
   a better `contention_event` test.
4. Add `crdb_internal.trace_id` builtin.

Release note: None

knz reviewed (commented):
Regarding the first commit: I am not convinceed by the unit test -- if we want to check that traces get collected across RPC boundaries, I'd expect a test that sets up a RPC server, sends an actuall RPC over a network connection, and asserts that the expected trace events are communicated.

In the 6th commit I dislike the approach you've taken, with more explanations about my view below.

(In contrast the 7th commit is super elegant and it was a relief to see that the way forward wasn't too painful)

The rest LGTM

> Reviewed 4 of 10 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, 2 of 2 files at r10, 10 of 10 files at r11, 8 of 8 files at r12, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-:-MT_uGlz2ugqTePRUPh_:b-6iwgc0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, and @tbg)

___
*[pkg/sql/crdb_internal.go, line 1181 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_wTq45e9Clbl78OVL:-MT_wTq52WGVZuR7T3KW:b-jdav6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/crdb_internal.go#L1181)):*
> ```Go
> 
> 				var numStructured int
> 				rec.Structured(func(any *ptypes.Any) {
> ```

Ouch.  I though the entire point of this vtable which leaves out the details of traces was to avoid iterating over all events while populating the table, so that the SQL client can apply a selection before we pay the cost of looking into a recording.

Can you not have a new built-in function which takes a trace/span ID as argument and returns a JSON blob with its properties, including whether it contains _any_ structured event as a boolean (and make the `rec.Structured` logic stop the iteration as soon as 1 is encountered).

For context, I think you are under-estimating the run-time and memory costs of crdb_internal tables generally and how much (ab)use they take by end-users and our support staff. This stuff needs to be programmed defensively with adequate explanatory comments about their costs in in the table's comment.
___
*[pkg/sql/txn_state.go, line 162 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_wE2W3CPUP2ePHVcz:-MT_wE2W3CPUP2ePHVd-:brdgeoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/txn_state.go#L162)):*
> ```Go
> 	var traceOpts []tracing.SpanOption
> 	if !tranCtx.execTestingKnobs.Pretend59315IsFixed {
> 		traceOpts = append(traceOpts, tracing.WithBypassRegistry())
> ```

Please add a comment to indicate that this conditional and the traceOpts can be removed entirely once the linked issue has been addressed.
___
*[pkg/sql/sem/builtins/builtins.go, line 3590 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_vgTEEFr5iyavcuDd:-MT_vgTFABnRhBClDQSO:byi9ne9) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/sem/builtins/builtins.go#L3590)):*
> ```Go
> 				sp := tracing.SpanFromContext(ctx.Context)
> 				if sp == nil {
> 					return nil, errors.New("no active trace span")
> ```

The proper way in SQL to indicate there's no value is to return NULL. I don't see a good reason to trash the SQL transaction in this case.
___
*[pkg/util/tracing/span.go, line 159 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_vGEW3k-qrI2Rm3YX:-MT_vGEW3k-qrI2Rm3YY:b-sa8til) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e97fefa70fb899d102070f527867d3b6273b019/pkg/util/tracing/span.go#L159)):*
> ```Go
> // recorded traces from other nodes.
> func (s *Span) ImportRemoteSpans(remoteSpans []tracingpb.RecordedSpan) error {
> 	if s.tracer.mode() == modeLegacy && s.crdb.recordingType() == RecordingOff {
> ```

Isn't this change needed at soon as the first commit? Especially if, as I suggested above, the first commit was extended to assert that trace events are in fact collected across remote RPCs?
___
*[pkg/util/tracing/tracer.go, line 299 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_tsco7mGUlkCNPxib:-MT_tscp-1Yu-GAhMnH4:b2mtaiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a932be4180c62972d3467208e874ce4aabd90f0d/pkg/util/tracing/tracer.go#L299)):*
> ```Go
> 	// also create a real child. Additionally, if the caller explicitly asked for
> 	// a real span they need to get one. In all other cases, a noop span will do.
> 	if !t.AlwaysTrace() &&
> ```

As a reviewer, I think that `!(t.AlwaysTrace() || opts.parentTraceID() != 0 || opts.ForceRealSpan)` gives me a more intuitive idea of what this condition is about (i.e. "if I don't have any reason to trace")


<!-- Sent from Reviewable.io -->

tbg committed:
tracing: simplify noop span conditional

Suggested by @knz.

Release note: None

tbg committed:
tracing: add real tests for the grpc interceptors

As was rightly pointed out by @knz during review in #59992, the tracing
interceptors were only tested indirectly.

In this commit, we add explicit tests that set up a gRPC server and
verify that for all combinations of unary/streaming the correct trace
spans get created.

Release note: None

tbg reviewed (commented):
TFTR!

> Regarding the first commit: I am not convinceed by the unit test -- if we want to check that traces get collected across RPC boundaries, I'd expect a test that sets up a RPC server, sends an actuall RPC over a network connection, and asserts that the expected trace events are communicated.

Good point - I added a comprehensive test for this in an additional commit.

> In the 6th commit I dislike the approach you've taken, with more explanations about my view below.

Fair point, I assume this is more a failure to communicate, see the inline discussion.

I added a few more commits - haven't done any squashing or rearranging to keep the review easier, PTAL.

> Dismissed @irfansharif from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-:-MTe_NrWA80HzFLganUo:b-fotngh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @angelapwen, @irfansharif, and @knz)

___
*[pkg/sql/crdb_internal.go, line 1181 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_wTq45e9Clbl78OVL:-MTeBlaq9oR11B2Tkus_:b-lacy99) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/crdb_internal.go#L1181)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ouch.  I though the entire point of this vtable which leaves out the details of traces was to avoid iterating over all events while populating the table, so that the SQL client can apply a selection before we pay the cost of looking into a recording.

Can you not have a new built-in function which takes a trace/span ID as argument and returns a JSON blob with its properties, including whether it contains _any_ structured event as a boolean (and make the `rec.Structured` logic stop the iteration as soon as 1 is encountered).

For context, I think you are under-estimating the run-time and memory costs of crdb_internal tables generally and how much (ab)use they take by end-users and our support staff. This stuff needs to be programmed defensively with adequate explanatory comments about their costs in in the table's comment.

</blockquote></details>

I agree with everything you're saying, and should have clarified how this code came to be (my bad). Essentially @angelapwen had trouble getting the structured data to show up in her attempt to add a built-in to render the payloads from a given span, and while investigating I found the random bits of breakage that caused this. I want this PR to give her a solid foundation to work from, and that meant adding at least a rudimentary test confirming that the payloads to indeed flow from where they originate to where they need to be. Long story short - I would like Angela (with your help) to tackle the shortcomings you're describing. I don't think `num_structured` should remain in the table once we have a way to ask for the payloads of a specified trace. This was just the easiest thing I could do.
___
*[pkg/sql/txn_state.go, line 162 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_wE2W3CPUP2ePHVcz:-MTeDAev7QtMJcGpUGoI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/txn_state.go#L162)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please add a comment to indicate that this conditional and the traceOpts can be removed entirely once the linked issue has been addressed.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 3590 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_vgTEEFr5iyavcuDd:-MTeDK65C1fC2R8ZUEYr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/sem/builtins/builtins.go#L3590)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The proper way in SQL to indicate there's no value is to return NULL. I don't see a good reason to trash the SQL transaction in this case.
</blockquote></details>

Done.
___
*[pkg/util/tracing/span.go, line 159 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_vGEW3k-qrI2Rm3YX:-MTe_FM_-r2s_dl4hPn4:b-3oni1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e97fefa70fb899d102070f527867d3b6273b019/pkg/util/tracing/span.go#L159)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Isn't this change needed at soon as the first commit? Especially if, as I suggested above, the first commit was extended to assert that trace events are in fact collected across remote RPCs?
</blockquote></details>

Yes, I will rearrange before the merge (when I squash the fixup commits).
___
*[pkg/util/tracing/tracer.go, line 299 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_tsco7mGUlkCNPxib:-MTeDl8nEJMa7jfXSl2Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a932be4180c62972d3467208e874ce4aabd90f0d/pkg/util/tracing/tracer.go#L299)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

As a reviewer, I think that `!(t.AlwaysTrace() || opts.parentTraceID() != 0 || opts.ForceRealSpan)` gives me a more intuitive idea of what this condition is about (i.e. "if I don't have any reason to trace")
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/sql/crdb_internal.go, line 1181 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MT_wTq45e9Clbl78OVL:-MTfsfRh1mnTo4d_LtHe:b4zsbf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/337d0cea8702c6ec8f368808bb53525984a02e4d/pkg/sql/crdb_internal.go#L1181)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I agree with everything you're saying, and should have clarified how this code came to be (my bad). Essentially @angelapwen had trouble getting the structured data to show up in her attempt to add a built-in to render the payloads from a given span, and while investigating I found the random bits of breakage that caused this. I want this PR to give her a solid foundation to work from, and that meant adding at least a rudimentary test confirming that the payloads to indeed flow from where they originate to where they need to be. Long story short - I would like Angela (with your help) to tackle the shortcomings you're describing. I don't think `num_structured` should remain in the table once we have a way to ask for the payloads of a specified trace. This was just the easiest thing I could do.
</blockquote></details>

Acknowledged — this can be removed now that #60616 successfully finds a contention event from a trace via payload. 
___
*[pkg/sql/crdb_internal.go, line 1187 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfu4hCBfBgU33e11qM:-MTfu4hCBfBgU33e11qN:by3ood8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/crdb_internal.go#L1187)):*
> ```Go
> 				if err := addRow(
> 					// TODO(angelapwen): we're casting uint64s to int64 here,
> 					// is that ok?
> ```

Hm, I see what you mean - there could be some casting problems if the unsigned int was large enough. I think the `DInt` needs to be a `int64` but we could potentially return an error if the ints are too large? 


<!-- Sent from Reviewable.io -->

angelapwen commented:
___
*[pkg/sql/sem/builtins/builtins.go, line 3595 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfwxg32ahsztlXjghd:-MTfwxg32ahsztlXjghe:b-b3fk08) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/sem/builtins/builtins.go#L3595)):*
> ```Go
> 			},
> 			Info: "Returns the current trace ID or an error if no trace is open.",
> 			// NB: possibly this is or could be made stable, but it's not worth it.
> ```

Nit: did you mean `possibly this is not`? Not sure what this means. 


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, 7 of 7 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-:-MTgAjR2C2mHjulakX7O:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/sql/crdb_internal.go, line 1187 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfu4hCBfBgU33e11qM:-MTgAm_tBxWUldq-YQid:bsaoqh4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/crdb_internal.go#L1187)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Hm, I see what you mean - there could be some casting problems if the unsigned int was large enough. I think the `DInt` needs to be a `int64` but we could potentially return an error if the ints are too large? 
</blockquote></details>

I think the conversion between uint and int in Go is lossless, i.e. it does not lose information. A large uint value will convert to a negative signed value, but the conversion remains bijective.

Since these are identifiers (and not quantities), we only care about the "identifying" nature of the values. 

However, I do agree it may be confusing for UX. @tbg was there any reason to make the trace/span IDs unsigned?
___
*[pkg/sql/sem/builtins/builtins.go, line 3595 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfwxg32ahsztlXjghd:-MTg9ViIDtdbu7pFWcIo:bopp7pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/sem/builtins/builtins.go#L3595)):*
<details><summary><i>Previously, angelapwen (Angela P Wen) wrote…</i></summary><blockquote>

Nit: did you mean `possibly this is not`? Not sure what this means. 
</blockquote></details>

I think the comment is correct, although tobias could have clarified "stable (i.e. non-volatile)".

(Angela: The volatibility attribute indicates whether the result of a built-in function is going to change throughout the execution of the statement.  Since our trace IDs are fixed for an entire SQL txn, at lease for an entire SQL stmt, I believe like Tobias that they are stable and thus this built-in can be marked as non-volatile, but I also agree it's a hassle to test / validate and so to be conservative we can keep this as is.)

@tbg I think also it's worth filing an issue for the optimizer team to look at. The queries we'd like to advertise to consume trace data will require SQL joins. If this builtin is marked volatile, the query plans will be more complex. We need some research to investigate the penalty and set the volatility properly if the penalty is high.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @angelapwen)

___
*[pkg/sql/crdb_internal.go, line 1187 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfu4hCBfBgU33e11qM:-MTjbtpI8MqpfWc8LqIA:bc6evnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/crdb_internal.go#L1187)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think the conversion between uint and int in Go is lossless, i.e. it does not lose information. A large uint value will convert to a negative signed value, but the conversion remains bijective.

Since these are identifiers (and not quantities), we only care about the "identifying" nature of the values. 

However, I do agree it may be confusing for UX. @tbg was there any reason to make the trace/span IDs unsigned?
</blockquote></details>

They have historically been unsigned, and I don't think there's a good reason. However, we can't just change it due to the necessary migration. Concretely, I think where the uintness matters is here:

https://github.com/cockroachdb/cockroach/blob/a1c3912438c68272a922d139abb42e6886190f58/pkg/util/tracing/tracer.go#L550-L551

and here:

https://github.com/cockroachdb/cockroach/blob/a1c3912438c68272a922d139abb42e6886190f58/pkg/util/tracing/tracer.go#L604-L615

So for 20.2 support, we have to keep it as a uint on the wire. Which should be fine! But honestly I'm not sure if there is enough of a reason here to make the change. `uint64` is a fine choice and nobody likes IDs that have signs. Can we make the ID columns `BIGINT`s instead? I do agree that we want the representation of IDs to match between internal and SQL, or we're going to be very confused once we walk further down the path in which trace messages are "stored" via log messages.

Filed https://github.com/cockroachdb/cockroach/issues/60668
___
*[pkg/sql/sem/builtins/builtins.go, line 3595 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfwxg32ahsztlXjghd:-MTje-ef29k_ORij8eG2:b-y1pnuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/sem/builtins/builtins.go#L3595)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think the comment is correct, although tobias could have clarified "stable (i.e. non-volatile)".

(Angela: The volatibility attribute indicates whether the result of a built-in function is going to change throughout the execution of the statement.  Since our trace IDs are fixed for an entire SQL txn, at lease for an entire SQL stmt, I believe like Tobias that they are stable and thus this built-in can be marked as non-volatile, but I also agree it's a hassle to test / validate and so to be conservative we can keep this as is.)

@tbg I think also it's worth filing an issue for the optimizer team to look at. The queries we'd like to advertise to consume trace data will require SQL joins. If this builtin is marked volatile, the query plans will be more complex. We need some research to investigate the penalty and set the volatility properly if the penalty is high.

</blockquote></details>

I think we should reach out when we have some actual queries, otherwise we're asking them in a vacuum.


<!-- Sent from Reviewable.io -->

angelapwen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/crdb_internal.go, line 1187 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfu4hCBfBgU33e11qM:-MTjkxsr2WYf7BcenzFZ:b8h5gf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/crdb_internal.go#L1187)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

They have historically been unsigned, and I don't think there's a good reason. However, we can't just change it due to the necessary migration. Concretely, I think where the uintness matters is here:

https://github.com/cockroachdb/cockroach/blob/a1c3912438c68272a922d139abb42e6886190f58/pkg/util/tracing/tracer.go#L550-L551

and here:

https://github.com/cockroachdb/cockroach/blob/a1c3912438c68272a922d139abb42e6886190f58/pkg/util/tracing/tracer.go#L604-L615

So for 20.2 support, we have to keep it as a uint on the wire. Which should be fine! But honestly I'm not sure if there is enough of a reason here to make the change. `uint64` is a fine choice and nobody likes IDs that have signs. Can we make the ID columns `BIGINT`s instead? I do agree that we want the representation of IDs to match between internal and SQL, or we're going to be very confused once we walk further down the path in which trace messages are "stored" via log messages.

Filed https://github.com/cockroachdb/cockroach/issues/60668
</blockquote></details>

Got it, we can move discussion to the issue filed! Good catch :smile: 
___
*[pkg/sql/sem/builtins/builtins.go, line 3595 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfwxg32ahsztlXjghd:-MTjlAbi2l5OKoPoBDkq:bpd336i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/sem/builtins/builtins.go#L3595)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think we should reach out when we have some actual queries, otherwise we're asking them in a vacuum.
</blockquote></details>

Thank you for the explanation of the volatility attribute @knz I got confused by the wording of the comment I think!


<!-- Sent from Reviewable.io -->

tbg committed:
sql: work around contention registry errors

See https://github.com/cockroachdb/cockroach/issues/60669.

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @angelapwen, @irfansharif, and @knz)

___
*[pkg/sql/crdb_internal.go, line 1187 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTfu4hCBfBgU33e11qM:-MTjw-QZ4MPfgyKluwTI:belcbqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f0fcb31654ab6019144a9bbf2367dd5f63bde87/pkg/sql/crdb_internal.go#L1187)):*
> Can we make the ID columns BIGINTs instead? 

That is not a thing. SQL BIGINT = signed int64

You may mean DECIMAL or BIT, but these come with obnoxious heap allocation overheads.


<!-- Sent from Reviewable.io -->

tbg committed:
testcluster: improve output of "no leaked trace spans" lint

Release note: None

tbg committed:
jobs: don't leak span when job is never started

TestProtectedTimestampsFailDueToLimits fails without this change.

I'm not in love with the way I fixed this - I think ideally a
`StartableJob` should have one method that is called from `Start` or
from `CleanupOnRollback`. As is, there are two cleanup paths, which
is a pattern that will rot again.

Filed #60671 to that effect.

Release note: None

tbg committed:
kvserver: disable a leaking trace span

See https://github.com/cockroachdb/cockroach/issues/60677.

The leak only becomes evident in the follow-up commit.

Release note: None

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just some nits, feel free to ignore

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-:-MTlc4MJCfqScSzuQ1f6:bzhopnq)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @angelapwen, @irfansharif, @knz, and @tbg)

___
*[pkg/sql/distsql_running.go, line 673 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTlZ9aj2sk6_nidYK8F:-MTlZ9akCAtGM0iJ5aHG:b-se64dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f32acedf59426e160b807d0125cf8ad4e72407b0/pkg/sql/distsql_running.go#L673)):*
> ```Go
> 					}
> 					if err := r.contentionRegistry.AddContentionEvent(ev); err != nil {
> 						// TODO(asubiotto): see #
> ```

Issue number missing
___
*[pkg/sql/logictest/logic.go, line 567 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTlbBikEOEdxzBYOeGV:-MTlbBikEOEdxzBYOeGW:bo3r2ea) ([raw file](https://github.com/cockroachdb/cockroach/blob/0518656bd8cfe1f27d5123199844df05935f03f1/pkg/sql/logictest/logic.go#L567)):*
> ```Go
> 	},
> 	{
> 		name:                "5node-pretend59315IsFixed",
> ```

[nit] `5node-pretend-59315-is-fixed`


<!-- Sent from Reviewable.io -->

tbg committed:
jobs: work around leaked trace span

See https://github.com/cockroachdb/cockroach/issues/60671.

Release note: None

tbg committed:
tracing: be more resilient to double-Finish

Our internal trace spans are fine, but net/trace tends to just crash
(not always, just sometimes). We don't *want* spans getting finished
twice, but it's hard to always avoid it, so add resilience.

This was prompted by a crash I saw in acceptance/version-upgrade which
very likely resulted from this hack:

https://github.com/cockroachdb/cockroach/issues/60671#issuecomment-780835710

Release note: None

tbg commented:
Not gonna waste my shot :notes: 
bors single=on p=999 r=knz,irfansharif,raduberinde

tbg commented:
bors single on
bors p=999
bors r=knz,irfansharif,raduberinde

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59992)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @angelapwen, @irfansharif, and @knz)

___
*[pkg/sql/logictest/logic.go, line 567 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTlbBikEOEdxzBYOeGV:-MTmADJE664zJGNchL-B:becbz4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0518656bd8cfe1f27d5123199844df05935f03f1/pkg/sql/logictest/logic.go#L567)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `5node-pretend-59315-is-fixed`
</blockquote></details>

I didn't do this, was too scared to eat another cycle


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/logictest/logic.go, line 567 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/59992#-MTlbBikEOEdxzBYOeGV:-MTmAa0s4RKCO2_76iWI:bq5nq6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0518656bd8cfe1f27d5123199844df05935f03f1/pkg/sql/logictest/logic.go#L567)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I didn't do this, was too scared to eat another cycle
</blockquote></details>

Heh, no problem. Hopefully it will be fixed soon so we won't have to pretend anymore :)


<!-- Sent from Reviewable.io -->
